Move nn.module state dict pre hook

zhaojuanmao opened the pull request:
Some modules like lazyModule may override '_save_to_state_dict()', in this case, pre_state_dict hook will not be called. So move the pre_state_dict hook out from '_save_to_state_dict()' to make sure the pre hook could be called

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98964

zhaojuanmao committed:
wq!

[ghstack-poisoned]

albanD reviewed (commented):
This would need test!

cc @rohan-varma

zhaojuanmao committed:
Update on "Move nn.module state dict pre hook "

Some modules like lazyModule may override '_save_to_state_dict()', in this case, pre_state_dict hook will not be called. So move the pre_state_dict hook out from '_save_to_state_dict()' to make sure the pre hook could be called




[ghstack-poisoned]

zhaojuanmao commented:
added a unit test for lazy module, without this change, lazy module unit test will fail

albanD commented on test/test_nn.py:
```diff
@@ -2425,10 +2425,35 @@ def my_state_dict_pre_hook(*args, **kwargs):
+        keep_var_setting = False
 
+        def my_state_dict_pre_hook(module, prefix, keep_vars):
+            nonlocal keep_var_setting
```
This is not needed if you only read the variable.

albanD commented on test/test_nn.py:
```diff
@@ -2425,10 +2425,35 @@ def my_state_dict_pre_hook(*args, **kwargs):
+        submodule.register_state_dict_pre_hook(my_state_dict_pre_hook)
+
+        def check_results(model):
+            nonlocal _state_dict_prefix, state_dict_pre_hook_count, keep_var_setting
```
Similar only needed for the things you actually modify

albanD reviewed (commented):
Only small nits on testing, sounds good otherwise.

zhaojuanmao committed:
Update on "Move nn.module state dict pre hook "

Some modules like lazyModule may override '_save_to_state_dict()', in this case, pre_state_dict hook will not be called. So move the pre_state_dict hook out from '_save_to_state_dict()' to make sure the pre hook could be called




[ghstack-poisoned]

albanD reviewed (approved):
Thanks!

zhaojuanmao committed:
Update on "Move nn.module state dict pre hook "

Some modules like lazyModule may override '_save_to_state_dict()', in this case, pre_state_dict hook will not be called. So move the pre_state_dict hook out from '_save_to_state_dict()' to make sure the pre hook could be called




[ghstack-poisoned]

zhaojuanmao committed:
Update on "Move nn.module state dict pre hook "

Some modules like lazyModule may override '_save_to_state_dict()', in this case, pre_state_dict hook will not be called. So move the pre_state_dict hook out from '_save_to_state_dict()' to make sure the pre hook could be called




[ghstack-poisoned]

zhaojuanmao committed:
Update on "Move nn.module state dict pre hook "

Some modules like lazyModule may override '_save_to_state_dict()', in this case, pre_state_dict hook will not be called. So move the pre_state_dict hook out from '_save_to_state_dict()' to make sure the pre hook could be called




[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot merge

zhaojuanmao commented:
@pytorchbot label "topic: not user facing"

zhaojuanmao commented:
@pytorchbot merge
